src/dlgConnectionProfiles.cpp

```diff
@@ -1662,6 +1676,156 @@ void dlgConnectionProfiles::fillout_form()
+    QPainter pt(&background);
+    pt.setCompositionMode(QPainter::CompositionMode_SourceOver);
+    pt.fillRect(QRect(0, 0, 120, 30), shade);
+    QPixmap pg(QStringLiteral(":/icons/mudlet_main_32px.png"));
+    pt.drawPixmap(QRect(5, 5, 20, 20), pg);
+
+    QFont _font;
+    QImage _pm(90, 30, QImage::Format_ARGB32_Premultiplied);
+    QPainter _pt(&_pm);
+    _pt.setCompositionMode(QPainter::CompositionMode_SourceOver);
+    int fs = 30;
+    for (; fs > 1; fs--) {
+        _pt.eraseRect(QRect(0, 0, 90, 30));
+        _pt.fillRect(QRect(0, 0, 90, 30), QColor(255, 0, 0, 10));
+        _font = QFont(QStringLiteral("DejaVu Sans"), fs, QFont::Normal);
+        _pt.setFont(_font);
+        QRect _r;
+        if ((i1 + i2 + i3 + i4 + i5 + i6) / 6 < 100) {
+            _pt.setPen(QColor(Qt::white));
+        } else {
+            _pt.setPen(QColor(Qt::black));
+        }
+        _pt.drawText(QRect(0, 0, 90, 30), Qt::AlignHCenter | Qt::AlignVCenter | Qt::TextWordWrap, profileName, &_r);
+        if (_r.width() <= 90 && _r.height() <= 30) {
+            break;
+        }
+    }
+    pt.setFont(_font);
+    QRect _r;
+    if ((i1 + i2 + i3 + i4 + i5 + i6) / 6 < 100) {
+        pt.setPen(QColor(Qt::white));
+    } else {
+        pt.setPen(QColor(Qt::black));
+    }
+        pt.drawText(QRect(30, 0, 90, 30), Qt::AlignHCenter | Qt::AlignVCenter | Qt::TextWordWrap, profileName, &_r);
+        pItem->setIcon(QIcon(background));
+}
+
+void dlgConnectionProfiles::slot_profile_menu(QPoint pos)
+{
+    QPoint globalPos = profiles_tree_widget->mapToGlobal(pos);
+    auto profileName = profiles_tree_widget->currentItem()->text();
+
+    QMenu menu;
+    if (hasCustomIcon(profileName)) {
+        menu.addAction(tr("Reset icon", "Reset the custom picture for this profile in the connection dialog and show the default one instead"), this, &dlgConnectionProfiles::slot_reset_custom_icon);
+    } else {
+        menu.addAction(QIcon(":/icons/mudlet_main_16px.png"), tr("Set custom icon", "Set a custom picture to show for the profile in the connection dialog"), this,  &dlgConnectionProfiles::slot_set_custom_icon);
+    }
+
+    menu.exec(globalPos);
+}
+
+void dlgConnectionProfiles::slot_set_custom_icon()
+{
+    auto profileName = profiles_tree_widget->currentItem()->text();
+
+    QString imageLocation = QFileDialog::getOpenFileName(this, tr("Select custom image for profile (should be 120x30)"),
+                                                    QStandardPaths::writableLocation(QStandardPaths::HomeLocation),
+                                                    tr("Images (%1)").arg(QStringLiteral("*.png *.gif *.jpg")));
+    if (imageLocation.isEmpty()) {
+        return;
+    }
+
+    bool success = mudlet::self()->setProfileIcon(profileName, imageLocation).first;
+    if (!success) {
+        return;
+    }
+
+    auto icon = QIcon(QPixmap(imageLocation).scaled(QSize(120, 30), Qt::IgnoreAspectRatio, Qt::SmoothTransformation).copy());
```

Comment:
Shouldn't this have happened already with `setProfileIcon`?

Reply:
`mudlet::setProfileIcon` - it just records the picture for use, and this _actually_ updates the connection window.

Reply:
But those 2 lines look to be the same as the last 2 lines of the now new `setCustomIcon`, which is called by `loadCustomProfile` which in turn is called by `setProfileItem`

Reply:
Yeah but this is `slot_set_custom_icon` where `setProfileItem` is not called?

Not called because if I were to use it here, it'd move the profile down to the bottom of the list - it blindly adds a profile, doesn't search if it already exists to find it.

Reply:
Oooh, I got confused with the different `setProfileIcon`s... The name is now defined in 3 different locations, which is... unfortunate?

Reply:
Yes, in different contexts it does a different thing. Not sure of better names.